rm-datasource/src/main/java/io/seata/rm/RMHandlerAT.java

```diff
@@ -40,43 +43,93 @@
 
     private static final int LIMIT_ROWS = 3000;
 
+    private final Map<String, Boolean> undoLogTableExistRecord = new ConcurrentHashMap<>();
+
     @Override
     public void handle(UndoLogDeleteRequest request) {
+        String resourceId = request.getResourceId();
         DataSourceManager dataSourceManager = (DataSourceManager)getResourceManager();
-        DataSourceProxy dataSourceProxy = dataSourceManager.get(request.getResourceId());
+        DataSourceProxy dataSourceProxy = dataSourceManager.get(resourceId);
         if (dataSourceProxy == null) {
-            LOGGER.warn("Failed to get dataSourceProxy for delete undolog on {}", request.getResourceId());
+            LOGGER.warn("Failed to get dataSourceProxy for delete undolog on {}", resourceId);
+            return;
+        }
+
+        boolean hasUndoLogTable = undoLogTableExistRecord.computeIfAbsent(resourceId, id -> checkUndoLogTableExist(dataSourceProxy));
```

Comment:
At the beginning, I didn't create this table, but after a while, I created this table. Is there any problem?

Reply:
if users don't have undo_log table at first, it means they must not using AT mode in their code, otherwise things would go wrong, right? Later if they decide to pick up AT mode and create undo_log table, they would need to modify their code and rebuild -> redeploy -> ...